rustls/tests/api.rs

```diff
@@ -3826,3 +3827,71 @@ fn test_server_rejects_clients_without_any_kx_group_overlap() {
         ))
     );
 }
+
+#[test]
+fn test_acceptor() {
+    let client_config = Arc::new(make_client_config(KeyType::ED25519));
+    let mut client = ClientConnection::new(client_config, dns_name("localhost")).unwrap();
+    let mut buf = Vec::new();
+    client.write_tls(&mut buf).unwrap();
+
+    let server_config = Arc::new(make_server_config(KeyType::ED25519));
+    let mut acceptor = rustls::Acceptor::new(None).unwrap();
+    assert!(acceptor.wants_read());
+    acceptor
+        .read_tls(&mut buf.as_slice())
+        .unwrap();
+    let accepted = acceptor.accept().unwrap().unwrap();
+    let ch = accepted.client_hello();
+    assert_eq!(ch.server_name(), Some("localhost"));
+
+    let key = server_config
+        .cert_resolver
+        .resolve(accepted.client_hello())
+        .unwrap();
```

Comment:
Feels a bit odd that we're (indirectly) passing the `cert_resolver` in in this case at all since cert resolution happens entirely in user code on this path, but I respect that having multiple config types would be an unjustified maintenance burden.

Reply:
It also raises a question on what would happen if the client hello indicates that a different `server_config` should be used. Would it be save to use that? Probably yes, since the `cert_resolver` could definitely be standalone.

It might be less ambiguous if `cert_resolver` would not be part of `server_config` at all. Is it there for backward compatibility reasons?

Reply:
I suppose we could pass the `ResolvesServerCert` in separately for `ServerConnection::new()` -- I'm happy to make that change.

>  It also raises a question on what would happen if the client hello indicates that a different server_config should be used.

I'm not sure what you mean by that? There is only one call to the cert resolver; either it happens while handling the `ClientHello` message internally, or it happens on the "outside" for users of the `Acceptor` API.

Reply:
> I'm not sure what you mean by that?

I just meant it is not extremely obvious if people can use a `server_config.cert_resolver` first individually, and then pass a completely different `server_config` to `into_connection`. But given that this field isn't really connected to the remains of `server_config`, its also not super confusing.

Reply:
I guess we could just go the other way and only pass `ServerConfig` into `into_connection()`, letting it call `config.cert_resolver.resolve()` for you. That might be more orthogonal at the cost of potentially more indirection.

Reply:
> I suppose we could pass the ResolvesServerCert in separately for ServerConnection::new() -- I'm happy to make that change.

That suits my tastes, at least.

> I guess we could just go the other way and only pass ServerConfig into into_connection(), letting it call config.cert_resolver.resolve() for you. That might be more orthogonal at the cost of potentially more indirection.

Then to e.g. do an async resolve, you'd have to do it separately and then pass in a trivial resolver, which is not the most discoverable pattern.

Reply:
>>  I guess we could just go the other way and only pass ServerConfig into into_connection(), letting it call config.cert_resolver.resolve() for you. That might be more orthogonal at the cost of potentially more indirection.

> Then to e.g. do an async resolve, you'd have to do it separately and then pass in a trivial resolver, which is not the most discoverable pattern.

Yes, you couldn't even pass the same resolver you use outside, because it can introduce a race. Imagine the following: We have a list of SNIs which are mapped to server configurations. Something along

```rust
/// Maps from SNI to the configuration we use
struct Configs(Arc<RwLock<HashMap<String, EndpointConfig>>>)

struct EndpointConfig{
    rustls_config: <ServerConfig>,
    cert: Arc<CertifiedKey>,
    /* other stuff */
}
```

We can now definitely implement `ResolvesServerCert ` for `Configs` and use it both for looking up the to be used `ServerConfig`, as well as store the resolver based on that in the passed `ServerConfig`.

If we do that, there is however a chance that the rustls lookup up the cert again from that map, and gets a different config than the application previously monitored - since the map can change concurrently.

To avoid that, the resolver in the `ServerConfig` would need to be one that returns a single static cert.

That's doable, but a bit on the weird end. Moving `ResolvesServerCert` completely outside sounds more reasonable.

**What could however be potentially done is simply storing a `CertifiedKey` now as a plain field in `ServerConfig`**

Since applications can build or hold various ServerConfigs and pass them based on the observed `ClientHello` that seems totally reasonable. It also makes sense in a way that `ResolvesServerCert` is the only callback in `ServerConfig`. Everything else are plain values too (like `ciphersuites`).

Reply:
For now, I did go with the solution where the resolver inside the config is used. The reason is that I feel it's not reasonable to expose the extra complexity in what I believe will still be the common case of instantiating `ServerConnection`s directly. Having to instantiate a `ResolvesServerCert` separately for all users adds substantial complexity to the API.

> Then to e.g. do an async resolve, you'd have to do it separately and then pass in a trivial resolver, which is not the most discoverable pattern.

Actually, you can now only instantiate a `ServerConfig` through a builder API, and the easiest way to set up certificate resolution in the builder API is to pass a single cert, so this doesn't seem like a problem to me.

> If we do that, there is however a chance that the rustls lookup up the cert again from that map, and gets a different config than the application previously monitored - since the map can change concurrently.

There is no looking up again -- we only look up the certificate once either way. Plus, if the caller supplies a `ResolvesServerCert`, the responsibility to get that right is on them.

Reply:
Actually we can resolve the server certificate twice in the case of a `HelloRetryRequest`, but the same thing could happen today.

Reply:
> For now, I did go with the solution where the resolver inside the config is used.

I'm not sure about this approach. It seems easy for a user to see that resolution still goes through a trait and assume that it therefore must be synchronous, without discovering the viable-but-elaborate workaround of implementing `ResolvesServerCert` for a newtype containing the e.g. asynchronously fetched results and ignoring the arguments to `resolve`.